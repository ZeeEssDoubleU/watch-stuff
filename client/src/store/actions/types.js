// api action types
export const YOUTUBE_LIBRARY_LOADED = "YOUTUBE_LIBRARY_LOADED";
