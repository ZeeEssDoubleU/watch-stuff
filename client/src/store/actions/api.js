// api action types
export const YOUTUBE_LIBRARY_LOADED = "YOUTUBE_LIBRARY_LOADED";

export const youtubeLibraryLoaded = () => ({
	type: YOUTUBE_LIBRARY_LOADED,
});