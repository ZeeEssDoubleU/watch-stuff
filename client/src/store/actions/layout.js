// api action types
export const types = {
	TOGGLE_SIDEBAR: "TOGGLE_SIDEBAR",
};

export const action_toggleSidebar = () => ({
	type: types.TOGGLE_SIDEBAR,
});
