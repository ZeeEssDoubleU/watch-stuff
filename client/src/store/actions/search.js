// video action types
export const types = {
	SEARCH_VIDEOS_REQUEST: "SEARCH_VIDEOS_REQUEST",
	SEARCH_VIDEOS_SUCCESS: "SEARCH_VIDEOS_SUCCESS",
	SEARCH_VIDEOS_FAILURE: "SEARCH_VIDEOS_FAILURE",
	SEARCH_VIDEOS_DETAILS_REQUEST: "SEARCH_VIDEOS_DETAILS_REQUEST",
	SEARCH_VIDEOS_DETAILS_SUCCESS: "SEARCH_VIDEOS_DETAILS_SUCCESS",
	SEARCH_VIDEOS_DETAILS_FAILURE: "SEARCH_VIDEOS_DETAILS_FAILURE",
};

export const action_fetchSearchVideos = {
	request: (searchQuery, nextPageToken, amount) => ({
		type: types.SEARCH_VIDEOS_REQUEST,
		payload: {
			searchQuery,
			nextPageToken,
			amount,
		},
	}),
	success: (response, searchQuery) => ({
		type: types.SEARCH_VIDEOS_SUCCESS,
		payload: {
			response,
			searchQuery,
		},
	}),
	failure: (response, searchQuery) => ({
		type: types.SEARCH_VIDEOS_FAILURE,
		payload: {
			response,
			searchQuery,
		},
	}),
};

export const action_fetchSearchVideosDetails = {
	request: searchQuery => ({
		type: types.SEARCH_VIDEOS_DETAILS_REQUEST,
		payload: { searchQuery },
	}),
	success: response => ({
		type: types.SEARCH_VIDEOS_DETAILS_SUCCESS,
		payload: response,
	}),
	failure: response => ({
		type: types.SEARCH_VIDEOS_DETAILS_FAILURE,
		payload: response,
	}),
};
