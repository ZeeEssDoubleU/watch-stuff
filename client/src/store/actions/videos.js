// video action types
export const types = {
	MOST_POPULAR_REQUEST: "MOST_POPULAR_REQUEST",
	MOST_POPULAR_SUCCESS: "MOST_POPULAR_SUCCESS",
	MOST_POPULAR_FAILURE: "MOST_POPULAR_FAILURE",
	VIDEO_CATEGORIES_REQUEST: "VIDEO_CATEGORIES_REQUEST",
	VIDEO_CATEGORIES_SUCCESS: "VIDEO_CATEGORIES_SUCCESS",
	VIDEO_CATEGORIES_FAILURE: "VIDEO_CATEGORIES_FAILURE",
	MOST_POPULAR_BY_CATEGORY_REQUEST: "MOST_POPULAR_BY_CATEGORY_REQUEST",
	MOST_POPULAR_BY_CATEGORY_SUCCESS: "MOST_POPULAR_BY_CATEGORY_SUCCESS",
	MOST_POPULAR_BY_CATEGORY_FAILURE: "MOST_POPULAR_BY_CATEGORY_FAILURE",
	WATCH_DETAILS_REQUEST: "WATCH_DETAILS_REQUEST",
	WATCH_DETAILS_SUCCESS: "WATCH_DETAILS_SUCCESS",
	WATCH_DETAILS_FAILURE: "WATCH_DETAILS_FAILURE",
};

export const action_fetchMostPopular = {
	request: (amount, loadDescription, nextPageToken) => ({
		type: types.MOST_POPULAR_REQUEST,
		payload: {
			amount,
			loadDescription,
			nextPageToken,
		},
	}),
	success: response => ({
		type: types.MOST_POPULAR_SUCCESS,
		payload: response,
	}),
	failure: response => ({
		type: types.MOST_POPULAR_FAILURE,
		payload: response,
	}),
};

export const action_fetchCategory = {
	request: () => ({
		type: types.VIDEO_CATEGORIES_REQUEST,
		payload: {},
	}),
	success: response => ({
		type: types.VIDEO_CATEGORIES_SUCCESS,
		payload: response,
	}),
	failure: response => ({
		type: types.VIDEO_CATEGORIES_FAILURE,
		payload: response,
	}),
};

export const action_fetchMostPopularByCategory = {
	request: categories => ({
		type: types.MOST_POPULAR_BY_CATEGORY_REQUEST,
		payload: { categories },
	}),
	success: (response, categories) => ({
		type: types.MOST_POPULAR_BY_CATEGORY_SUCCESS,
		payload: {
			response,
			categories,
		},
	}),
	failure: response => ({
		type: types.MOST_POPULAR_BY_CATEGORY_FAILURE,
		payload: response,
	}),
};

export const action_fetchWatchDetails = {
	request: videoId => ({
		type: types.WATCH_DETAILS_REQUEST,
		payload: { videoId },
	}),
	success: response => ({
		type: types.WATCH_DETAILS_SUCCESS,
		payload: {
			response,
		},
	}),
	failure: response => ({
		type: types.WATCH_DETAILS_FAILURE,
		payload: response,
	}),
};
