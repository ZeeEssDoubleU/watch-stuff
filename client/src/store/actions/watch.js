// video action types
export const types = {
	WATCH_DETAILS_REQUEST: "WATCH_DETAILS_REQUEST",
	WATCH_DETAILS_SUCCESS: "WATCH_DETAILS_SUCCESS",
	WATCH_DETAILS_FAILURE: "WATCH_DETAILS_FAILURE",
	RELATED_VIDEOS_REQUEST: "RELATED_VIDEOS_REQUEST",
	RELATED_VIDEOS_SUCCESS: "RELATED_VIDEOS_SUCCESS",
	RELATED_VIDEOS_FAILURE: "RELATED_VIDEOS_FAILURE",
	RELATED_VIDEO_DETAILS_REQUEST: "RELATED_VIDEO_DETAILS_REQUEST",
	RELATED_VIDEO_DETAILS_SUCCESS: "RELATED_VIDEO_DETAILS_SUCCESS",
	RELATED_VIDEO_DETAILS_FAILURE: "RELATED_VIDEO_DETAILS_FAILURE",
	CHANNEL_DETAILS_REQUEST: "CHANNEL_DETAILS_REQUEST",
	CHANNEL_DETAILS_SUCCESS: "CHANNEL_DETAILS_SUCCESS",
	CHANNEL_DETAILS_FAILURE: "CHANNEL_DETAILS_FAILURE",
	COMMENTS_REQUEST: "COMMENTS_REQUEST",
	COMMENTS_SUCCESS: "COMMENTS_SUCCESS",
	COMMENTS_FAILURE: "COMMENTS_FAILURE",
};

export const action_fetchWatchDetails = {
	request: videoId => ({
		type: types.WATCH_DETAILS_REQUEST,
		payload: { videoId },
	}),
	success: response => ({
		type: types.WATCH_DETAILS_SUCCESS,
		payload: response,
	}),
	failure: response => ({
		type: types.WATCH_DETAILS_FAILURE,
		payload: response,
	}),
};

export const action_fetchRelatedVideos = {
	request: (videoId, amount) => ({
		type: types.RELATED_VIDEOS_REQUEST,
		payload: {
			videoId,
			amount,
		},
	}),
	success: (response, videoId) => ({
		type: types.RELATED_VIDEOS_SUCCESS,
		payload: {
			response,
			videoId,
		},
	}),
	failure: response => ({
		type: types.RELATED_VIDEOS_FAILURE,
		payload: response,
	}),
};

export const action_fetchRelatedVideoDetails = {
	request: videoIds => ({
		type: types.RELATED_VIDEO_DETAILS_REQUEST,
		payload: { videoIds },
	}),
	success: response => ({
		type: types.RELATED_VIDEO_DETAILS_SUCCESS,
		payload: response,
	}),
	failure: response => ({
		type: types.RELATED_VIDEO_DETAILS_FAILURE,
		payload: response,
	}),
};

export const action_fetchChannelDetails = {
	request: channelId => ({
		type: types.CHANNEL_DETAILS_REQUEST,
		payload: { channelId },
	}),
	success: response => ({
		type: types.CHANNEL_DETAILS_SUCCESS,
		payload: response,
	}),
	failure: response => ({
		type: types.CHANNEL_DETAILS_FAILURE,
		payload: response,
	}),
};

export const action_fetchComments = {
	request: (videoId, nextPageToken) => ({
		type: types.COMMENTS_REQUEST,
		payload: {
			videoId,
			nextPageToken,
		},
	}),
	success: (response, videoId) => ({
		type: types.COMMENTS_SUCCESS,
		payload: {
			response,
			videoId,
		},
	}),
	failure: response => ({
		type: types.COMMENTS_FAILURE,
		payload: response,
	}),
};
