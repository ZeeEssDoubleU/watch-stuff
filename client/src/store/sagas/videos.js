import { call, fork, take, takeEvery } from "redux-saga/effects";
import * as youtubeApi from "../api/youtube-api";
import * as videoActions from "../actions/videos";
import { fetchEntity } from "./index";

export function* watchMostPopularVideos() {
	yield takeEvery(videoActions.MOST_POPULAR_REQUEST, fetchMostPopularVideos);
}
export function* fetchMostPopularVideos(action) {
	console.log("ACTION", action);
	const request = () =>
		youtubeApi.buildMostPopularVideosRequest(
			action.payload.amount,
			action.payload.loadDescription,
			action.payload.nextPageToken,
      );
	yield fetchEntity(request, videoActions.fetchMostPopularVideos);
}