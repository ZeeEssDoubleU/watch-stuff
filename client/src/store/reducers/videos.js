import { createSelector } from "reselect";
import * as videoActions from "../actions/videos";
import * as watchActions from "../actions/watch";

const initialState = {
	categories: {},
	mostPopular: {},
	byId: {},
	byCategory: {},
	relatedVideos: {},
};

//***************
// VIDEO REDUCERS
//***************

const reducer_videos = (state = initialState, action) => {
	switch (action.type) {
		case videoActions.types.MOST_POPULAR_SUCCESS:
			return reducer_fetchMostPopular(action.payload, state);
		case videoActions.types.VIDEO_CATEGORIES_SUCCESS:
			return reducer_fetchVideoCategories(action.payload, state);
		case videoActions.types.MOST_POPULAR_BY_CATEGORY_SUCCESS:
			return reducer_fetchMostPopularByCategory(action.payload, state);
		case watchActions.types.WATCH_DETAILS_SUCCESS:
			return reducer_fetchWatchDetails(action.payload, state);
		case watchActions.types.RELATED_VIDEOS_SUCCESS:
			return reducer_fetchRelatedVideos(action.payload, state);
		case watchActions.types.RELATED_VIDEO_DETAILS_SUCCESS:
			return reducer_fetchRelatedVideoDetails(action.payload, state);
		default:
			return state;
	}
};

export default reducer_videos;

const reducer_fetchMostPopular = (payload, state) => {
	const videoMap = payload.items.reduce((obj, item) => {
		obj[item.id] = item;
		return obj;
	}, {});

	let itemIds = Object.keys(videoMap);
	console.log("PAYLOAD - MOST POPULAR VIDEOS", payload);
	console.log("MAP - MOST POPULAR VIDEOS BY ID", videoMap);

	// if prevPageToken exists, previous vids (items) have already been fetched from endpoint
	// combine previous vids into mostPopular itemIds (video ids)
	if (payload.hasOwnProperty("prevPageToken") && state.mostPopular) {
		itemIds = [...state.mostPopular.itemIds, ...itemIds];
	}

	const mostPopular = {
		totalResults: payload.pageInfo.totalResults,
		nextPageToken: payload.nextPageToken,
		itemIds,
	};

	// combine previous vids into state (same as above)
	return {
		...state,
		mostPopular,
		byId: {
			...state.byId,
			...videoMap,
		},
	};
};

const reducer_fetchVideoCategories = (payload, state) => {
	const categoryMap = payload.items.reduce((obj, item) => {
		obj[item.id] = item.snippet.title;
		return obj;
	}, {});

	console.log("PAYLOAD - VIDEO CATEGORIES", payload);
	console.log("MAP - VIDEO CATEGORIES", categoryMap);

	// combine previous vids into state (same as above)
	return {
		...state,
		categories: categoryMap,
	};
};

const reducer_fetchMostPopularByCategory = (payload, state) => {
	const { categories, response } = payload;
	const byIdMap = {};
	const byCategoryMap = {};

	response.forEach((category, index) => {
		if (!category.hasOwnProperty("errors")) {
			if (category.items.length > 0) {
				// map categories to byCategoryMap
				byCategoryMap[categories[index]] = {
					nextPageToken: category.nextPageToken,
					pageInfo: category.pageInfo,
					// map only video ids to category items.  Rest of video data exists in byId
					items: category.items.map(item => item.id),
				};
				// map category items (using item.id) to byIdMap
				category.items.forEach(item => (byIdMap[item.id] = item));
			}
		}
	});

	console.log("PAYLOAD - MOST POPULAR VIDEOS BY CATEGORY", payload);
	console.log("MAP - MOST POPULAR VIDEOS BY CATEGORY", byCategoryMap);
	console.log("MAP - CATEGORY VIDEOS BY ID", byIdMap);

	return {
		...state,
		byCategory: {
			...state.byCategory,
			...byCategoryMap,
		},
		byId: {
			...state.byId,
			...byIdMap,
		},
	};
};

//***************
// WATCH REDUCERS
//***************

// this reducer is mirrors in watch.js as reducer_fetchWatchDetails
const reducer_fetchWatchDetails = (payload, state) => {
	console.log("PAYLOAD - FETCH WATCH DETAILS", payload);
	const details = payload.items[0];
	const videoId = details.id;

	// add watch video details to byId lookup table
	return {
		...state,
		byId: {
			...state.byId,
			[videoId]: details,
		},
	};
};

const reducer_fetchRelatedVideos = (payload, state) => {
	const response = payload.response;
	const videoIds = response.items.map(item => item.id.videoId);

	console.log("PAYLOAD - RELATED VIDEOS", response);
	console.log("MAP - RELATED VIDEO IDS", videoIds);

	const relatedVideos = {
		totalResults: response.pageInfo.totalResults,
		nextPageToken: response.nextPageToken,
		videoIds,
	};

	// place related video ids into lookup table based on original videoId
	return {
		...state,
		relatedVideos: {
			...state.relatedVideos,
			[payload.videoId]: relatedVideos,
		},
	};
};

const reducer_fetchRelatedVideoDetails = (payload, state) => {
	const videoMap = payload.reduce((obj, response) => {
		const video = response.items ? response.items[0] : null;
		if (!video) return obj;
		obj[video.id] = video;
		return obj;
	}, {});
	console.log("PAYLOAD - FETCH RELATED VIDEO DETAILS", payload);
	console.log("MAP - RELATED VIDEOS BY ID", videoMap);

	// add related video details to byId lookup table
	return {
		...state,
		byId: {
			...state.byId,
			...videoMap,
		},
	};
};

//***************
// SELECTORS
//***************
export const selector_mostPopularVideos = createSelector(
	state => state.videosState.byId,
	state => state.videosState.mostPopular,
	(videosById, mostPopular) => {
		// if no mostPopular items exist, return []
		if (!mostPopular || !mostPopular.itemIds) {
			return [];
		}
		// map mostPopular items to an array, using mostPopular itemIds in byId lookup table
		return mostPopular.itemIds.map(id => videosById[id]);
	},
);

export const selector_videoCategories = createSelector(
	state => state.videosState.categories,
	categories => Object.keys(categories),
);

export const selector_videoCategoriesLoaded = createSelector(
	state => state.videosState.categories,
	categories => Object.keys(categories).length > 0,
);

export const selector_mostPopularVideosByCategory = createSelector(
	state => state.videosState.categories,
	state => state.videosState.byCategory,
	state => state.videosState.byId,
	(videoCategories, videosByCategory, videosById) => {
		const byCategory = {};
		const categoryIds = Object.keys(videosByCategory);
		categoryIds.forEach(categoryId => {
			const categoryName = videoCategories[categoryId];
			const videoIds = videosByCategory[categoryId].items;
			byCategory[categoryName] = videoIds.map(
				videoId => videosById[videoId],
			);
		});
		return byCategory;
	},
);

export const selector_mostPopularVideosByCategoryLoaded = createSelector(
	state => state.videosState.byCategory,
	byCategory => Object.keys(byCategory).length > 0,
);

export const selector_mostPopularVideosByCategoryLength = createSelector(
	state => state.videosState.byCategory,
	byCategory => Object.keys(byCategory).length,
);

// this selector is mirrored in watch.js as selector_watchDetails
export const selector_videoById = (state, videoId) =>
	state.videosState.byId[videoId];

// export const selector_relatedVideos = createSelector(

// )
