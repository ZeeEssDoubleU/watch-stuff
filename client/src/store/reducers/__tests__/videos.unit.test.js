import reducer_videos from "../videos";
import * as videoActions from "../../actions/videos";
import mostPopularResponse from "./responses/MOST_POPULAR_SUCCESS.json";
import mostPopularResponse_withPrevPageToken from "./responses/MOST_POPULAR_SUCCESS_withPrevPageToken.json";
import mostPopularSuccessState_prevState from "./states/MOST_POPULAR_SUCCESS_prevState.json";
import mostPopularSuccessState from "./states/MOST_POPULAR_SUCCESS.json";
import mostPopularSuccessState_withPrevPageToken from "./states/MOST_POPULAR_SUCCESS_withPrevPageToken.json";

const initialState = {
	mostPopular: {},
	categories: {},
	byId: {},
	byCategory: {},
};

describe("videos reducer", () => {
	test("test unused action type with default initial state", () => {
		const startState = undefined;
		const action = { type: "UNUSED_ACTION_TYPE" };
		const expectedEndState = { ...initialState };
		expect(reducer_videos(startState, action)).toEqual(expectedEndState);
		expect(expectedEndState).toMatchSnapshot();
	});

	test("test with MOST_POPULAR_SUCCESS action", () => {
		const startState = { ...initialState };
		const action = {
			type: videoActions.types.MOST_POPULAR_SUCCESS,
			payload: mostPopularResponse,
		};
		const expectedEndState = {
			...startState,
			...mostPopularSuccessState,
		};
		expect(reducer_videos(startState, action)).toEqual(expectedEndState);
		expect(expectedEndState).toMatchSnapshot();
	});

	test("test for idempotence with MOST_POPULAR_SUCCESS action", () => {
		const startState = mostPopularSuccessState_prevState;
		const action = {
			type: videoActions.types.MOST_POPULAR_SUCCESS,
			payload: mostPopularResponse,
		};
		// should keep and add to previous byId data, but not keep previous omstPopular
		const expectedEndState = {
			...startState,
			mostPopular: {
				...mostPopularSuccessState.mostPopular,
			},
			byId: {
				...startState.byId,
				...mostPopularSuccessState.byId,
			},
		};
		expect(reducer_videos(startState, action)).toEqual(expectedEndState);
		expect(expectedEndState).toMatchSnapshot();
	});

	test("test for idempotence with MOST_POPULAR_SUCCESS action with prevPageToken", () => {
		const startState = mostPopularSuccessState_prevState;
		const action = {
			type: videoActions.types.MOST_POPULAR_SUCCESS,
			payload: mostPopularResponse_withPrevPageToken,
		};
		// should keep and add to previous byId and mostPopular data
		const expectedEndState = {
			...startState,
			mostPopular: {
				...mostPopularSuccessState.mostPopular,
				itemIds: [
					...startState.mostPopular.itemIds,
					...mostPopularSuccessState.mostPopular.itemIds,
				],
			},
			byId: {
				...startState.byId,
				...mostPopularSuccessState.byId,
			},
		};
		expect(reducer_videos(startState, action)).toEqual(expectedEndState);
		expect(expectedEndState).toMatchSnapshot();
	});
});
