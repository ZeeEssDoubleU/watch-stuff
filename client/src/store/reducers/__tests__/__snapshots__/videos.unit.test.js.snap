// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`videos reducer test for idempotence with MOST_POPULAR_SUCCESS action 1`] = `
Object {
  "byId": Object {
    "eSLe4HuKuK0": Object {
      "contentDetails": Object {
        "duration": "PT2M",
      },
      "id": "eSLe4HuKuK0",
      "snippet": Object {
        "channelId": "UC_976xMxPgzIa290Hqtk-9g",
        "channelTitle": "Walt Disney Animation Studios",
        "localized": Object {
          "title": "Frozen 2 | Official Teaser Trailer",
        },
        "publishedAt": "2019-02-13T14:00:05.000Z",
        "thumbnails": Object {
          "medium": Object {
            "height": 180,
            "url": "https://i.ytimg.com/vi/eSLe4HuKuK0/mqdefault.jpg",
            "width": 320,
          },
        },
        "title": "Frozen 2 | Official Teaser Trailer",
      },
      "statistics": Object {
        "viewCount": "5486363",
      },
    },
    "hhzYbExfVIY": Object {
      "contentDetails": Object {
        "duration": "PT3M25S",
      },
      "id": "hhzYbExfVIY",
      "snippet": Object {
        "channelId": "UCM1gEqLPsgej7ZKzu9w8v7A",
        "channelTitle": "LilBabyVEVO",
        "localized": Object {
          "title": "Lil Baby, Gunna - Close Friends (Official Music Video)",
        },
        "publishedAt": "2019-02-13T15:00:06.000Z",
        "thumbnails": Object {
          "medium": Object {
            "height": 180,
            "url": "https://i.ytimg.com/vi/hhzYbExfVIY/mqdefault.jpg",
            "width": 320,
          },
        },
        "title": "Lil Baby, Gunna - Close Friends (Official Music Video)",
      },
      "statistics": Object {
        "viewCount": "929603",
      },
    },
  },
  "mostPopular": Object {
    "itemIds": Array [
      "eSLe4HuKuK0",
      "hhzYbExfVIY",
    ],
    "nextPageToken": "CAwQAA",
    "totalResults": 200,
  },
}
`;

exports[`videos reducer test for idempotence with MOST_POPULAR_SUCCESS action with prevPageToken 1`] = `
Object {
  "byId": Object {
    "aVt5uHuKuK0": Object {
      "contentDetails": Object {
        "duration": "PT2M",
      },
      "id": "aVt5uHuKuK0",
      "snippet": Object {
        "channelId": "AB_f798edPgzIa290Hqtk-9g",
        "channelTitle": "Gargantuan Entertainment",
        "localized": Object {
          "title": "Yo Mama | Official 'She So Fat' Trailer",
        },
        "publishedAt": "2019-02-13T14:00:05.000Z",
        "thumbnails": Object {
          "medium": Object {
            "height": 180,
            "url": "https://i.ytimg.com/vi/aVt5uHuKuK0/mqdefault.jpg",
            "width": 320,
          },
        },
        "title": "Yo Mama | Official 'She So Fat' Trailer",
      },
      "statistics": Object {
        "viewCount": "4032498342",
      },
    },
    "eSLe4HuKuK0": Object {
      "contentDetails": Object {
        "duration": "PT2M",
      },
      "id": "eSLe4HuKuK0",
      "snippet": Object {
        "channelId": "UC_976xMxPgzIa290Hqtk-9g",
        "channelTitle": "Walt Disney Animation Studios",
        "localized": Object {
          "title": "Frozen 2 | Official Teaser Trailer",
        },
        "publishedAt": "2019-02-13T14:00:05.000Z",
        "thumbnails": Object {
          "medium": Object {
            "height": 180,
            "url": "https://i.ytimg.com/vi/eSLe4HuKuK0/mqdefault.jpg",
            "width": 320,
          },
        },
        "title": "Frozen 2 | Official Teaser Trailer",
      },
      "statistics": Object {
        "viewCount": "5486363",
      },
    },
    "hhzYbExfVIY": Object {
      "contentDetails": Object {
        "duration": "PT3M25S",
      },
      "id": "hhzYbExfVIY",
      "snippet": Object {
        "channelId": "UCM1gEqLPsgej7ZKzu9w8v7A",
        "channelTitle": "LilBabyVEVO",
        "localized": Object {
          "title": "Lil Baby, Gunna - Close Friends (Official Music Video)",
        },
        "publishedAt": "2019-02-13T15:00:06.000Z",
        "thumbnails": Object {
          "medium": Object {
            "height": 180,
            "url": "https://i.ytimg.com/vi/hhzYbExfVIY/mqdefault.jpg",
            "width": 320,
          },
        },
        "title": "Lil Baby, Gunna - Close Friends (Official Music Video)",
      },
      "statistics": Object {
        "viewCount": "929603",
      },
    },
  },
  "mostPopular": Object {
    "itemIds": Array [
      "aVt5uHuKuK0",
      "eSLe4HuKuK0",
      "hhzYbExfVIY",
    ],
    "nextPageToken": "CAwQAA",
    "totalResults": 200,
  },
}
`;
